src/CompiledMatcher.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of composer/semver.
+ *
+ * (c) Composer <https://github.com/composer>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace Composer\Semver;
+
+use Composer\Semver\Constraint\ConstraintInterface;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class CompiledMatcher
+{
+    private $cacheDir;
+
+    private $checked;
+
+    public function __construct($cacheDir = null)
+    {
+        $this->cacheDir = $cacheDir ? $cacheDir : \sys_get_temp_dir();
+    }
+
+    /**
+     * Evaluates the expression: $constraint match $operator $version
+     *
+     * @param ConstraintInterface $constraint
+     * @param int                 $operator
+     * @param string              $version
+     *
+     * @return mixed
+     */
+    public function match(ConstraintInterface $constraint, $operator, $version)
+    {
+        $cacheKey = $operator.$constraint;
+        if (!isset($this->checked[$cacheKey])) {
+            $sha = \sha1($cacheKey);
+            $file = $this->cacheDir.'/constraint_'.$sha;
```

Comment:
shouldn't the cache dir or the cache file contain some kind of reference to the version of the compilation logic ? With the current approach, if we fix a bug in the compilation logic, all users of composer will have to delete the temporary cache files to be sure that the old buggy functions are not used.

Reply:
Great idea!.

Do we have an easy way to find the current version of the installed lib?

Reply:
`Composer\InstalledVersions` in Composer 2 :smile: 

But we might not need to bust the cache on each version. We could have an internal constant storing an "optimization version", that we bump when we change the optimization logic.

Reply:
not needed anymore, switched to `eval` impl